keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,631 @@
+- Backend pods for a service can be dual stack.
+- Endpoints for a dual-stack backend pod will be represented as a dual-stack address pair (i.e. 1 IPv4/IPv6 endpoint per backend pod, rather than 2 single-family endpoints per backend pod)
+- Kube-proxy iptables mode needs to drive iptables and ip6tables in parallel. This is required, even though service IP support is single-family, so that Kubernetes services can be exposed to clients external to the cluster via both IPv4 and IPv6. Support includes:
+  - Service IPs: Single family support (either all IPv4 or all IPv6 service IPs in a cluster)
+  - NodePort: Support listening on both IPv4 and IPv6 addresses
+  - ExternalIPs: Can be IPv4 or IPv6
+- Kube-proxy IPVS mode will support dual-stack functionality similar to kube-proxy iptables mode as described above. IPVS kube-router support for dual stack, on the other hand, is considered outside of the scope of this proposal.
+- For health/liveness/readiness probe support, the default behavior will not change and an additional optional field would be added to the pod specification and is respected by kubelet. This will allow application developers to select a preferred IP family to use for implementing probes on dual-stack pods.
+- The pod status API changes will include a per-IP string map for arbitrary annotations, as a placeholder for future Kubernetes enhancements. This mapping is not required for this dual-stack design, but will allow future annotations, e.g. allowing a CNI network plugin to indicate to which network a given IP address applies. The approriate hooks will be provided to enable CRI/CNI to provide these details.
+- Kubectl commands and output displays will need to be modified for dual-stack.
+- Kubeadm support will need to be added to enable spin-up of dual-stack clusters. Kubeadm support is required for implementing dual-stack continuous integration (CI) tests.
+- New e2e test cases will need to be added to test parallel IPv4/IPv6 connectivity between pods, nodes, and services.
+
+### Implementation Plan
+
+Given the scope of this enhancement, it has been suggested that we break the implementation into descrete phases that may be spread out over the span of many release cycles. The phases are as follows:
+
+- pod-to-pod dual-stack routing and non-vip services
+- Single-family services
+- Multi-family services
+- External dependencies, eg. cloud-provide, CNI, CRI, CoreDNS etc...
+
+### Awareness of Multiple IPs per Pod
+
+Since Kubernetes Version 1.9, Kubernetes users have had the capability to use dual-stack-capable CNI network plugins (e.g. Bridge + Host Local, Calico, etc.), using the 
+[0.3.1 version of the CNI Networking Plugin API](https://github.com/containernetworking/cni/blob/spec-v0.3.1/SPEC.md), to configure multiple IPv4/IPv6 addresses on pods. However, Kubernetes currently captures and uses only IP address from the pod's main interface.
```

Comment:
To be fair it's really the CRI that does this, not Kubernetes itself. So dockershim/CRI-O/etc are the things that need to transmit the multiple IPs back to kubelet via the CRI API.

Reply:
Since dockershim is in-tree, the feature is still depending on CRI/CNI to support multiple IPs.

Reply:
It seems the goal is to move it out-of-tree in 1.18 https://github.com/kubernetes/enhancements/pull/866, @dcbw do you think that this KEP can be affected by the deprecation plan  https://github.com/kubernetes/enhancements/pull/866/files#diff-8e35fdc72dca61e8e5f9c0ebe13ffd33R95 ?